prog/vm/nexus.rb

```diff
@@ -176,16 +176,20 @@ def allocation_dataset
     version_qualifier = requires_bdev_ubi ? "like '%ubi%'" : "not like '%ubi%'"
     spdk_where_clause = "AND id in (select vm_host_id from spdk_installation where version #{version_qualifier} and allocation_weight > 0)"
     DB[<<SQL, vm.cores, vm.mem_gib_ratio, vm.mem_gib, vm_storage_size_gib, vm.location, vm.arch]
-SELECT *, vm_host.total_mem_gib / vm_host.total_cores AS mem_ratio
-FROM vm_host
-WHERE vm_host.used_cores + ? < least(vm_host.total_cores, vm_host.total_mem_gib / ?)
+WITH vm_stats AS (SELECT vm_host.id host_id, COALESCE(sum(cores),0) AS cores FROM vm_host LEFT OUTER JOIN vm ON vm_host.id=vm.vm_host_id GROUP BY vm_host.id),
+spdk_stats AS (SELECT vm_host.id host_id, COALESCE(sum(core_count),0) AS cores FROM vm_host LEFT OUTER JOIN spdk_installation s ON vm_host.id=s.vm_host_id GROUP BY vm_host.id)
+SELECT vm_host.*, vm_host.total_mem_gib / vm_host.total_cores AS mem_ratio, vm_stats.cores + spdk_stats.cores AS used_core_count
+FROM vm_host, vm_stats, spdk_stats
+WHERE
+vm_host.id=vm_stats.host_id AND vm_host.id=spdk_stats.host_id AND
+vm_stats.cores + spdk_stats.cores + ? < least(vm_host.total_cores, vm_host.total_mem_gib / ?)
```

Comment:
This is pretty gnarly, what's going on here, generally speaking?

Reply:
I create a CTE `vm_stats` with 2 columns (host_id, cores), and another CTE `spdk_stats` with 2 columns (host_id, cores), then join them with the vm_host, so I can filter using `vm_stats.cores + spdk_stats.cores` (which is the up-to-date value of what we used to call used_cores)

One other alternative is to create a view and use it here instead of putting the SQL here, with the advantage that one can also use it for debugging.

Reply:
Maybe it's worth a comment. and indentation, to guide the eye.

Reply:
I am wondering if we should do this with `SELECT FOR UPDATE`. It would be a very strict lock but with this new form of the allocator, I feel like we'll have synchronisation issues.